### PR TITLE
Consistent way of checking Airflow version in providers

### PR DESCRIPTION
**Pull request by potiuk**
This PR introduces consistent way of checking version of Airflow
by Airflow providers. So far there were about 6 different ways on
how Providers checked for Airflow version - this PR aims to unify
this approach for now and in the future - at least until minimum
version of Airflow set to 2.11 where we are likely to introduce
a simpler check via https://github.com/apache/airflow/pull/44607. Until then all providers are going
to have `version_references.py` module copied in their sources
that they will be importing the constants from.

This PR also adds pre-commit that checks if the
``version_compat.py`` module is imported from local package copy
or maybe from another provider or test code - both causing
unneeded dependencies from the provider - to another package or
to test code respectively.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: providers/src/airflow/providers/amazon/aws/assets/s3.py [hunk 1]
```diff
@@ -28,17 +29,10 @@
         Dataset as OpenLineageDataset,
     )
 else:
-    # TODO: Remove this try-exception block after bumping common provider to 1.3.0
-    # This is due to common provider AssetDetails import error handling
```
Comment (Lee-W) [inline providers/src/airflow/providers/amazon/aws/assets/s3.py]: This is probably still needed

File: providers/src/airflow/providers/common/io/assets/file.py [hunk 2]
```diff
@@ -19,13 +19,9 @@
 import urllib.parse
 from typing import TYPE_CHECKING
 
-from packaging.version import Version
+from airflow.providers.common.io import PROVIDERS_COMMON_IO_AIRFLOW_IS_3_0_PLUS
 
-from airflow import __version__ as AIRFLOW_VERSION
-
-# TODO: Remove version check block after bumping common provider to 1.3.0
```
Comment (Lee-W) [inline providers/src/airflow/providers/common/io/assets/file.py:26]: This is probably still needed

File: providers/src/airflow/providers/google/assets/gcs.py [hunk 3]
```diff
@@ -26,17 +27,10 @@
     from airflow.providers.common.compat.assets import Asset
     from airflow.providers.common.compat.openlineage.facet import Dataset as OpenLineageDataset
 else:
-    # TODO: Remove this try-exception block after bumping common provider to 1.3.0
```
Comment (Lee-W) [inline providers/src/airflow/providers/google/assets/gcs.py:29]: This TODO is probably still needed

File: providers/src/airflow/providers/openlineage/utils/utils.py [hunk 4]
```diff
@@ -74,17 +78,14 @@
     from airflow.providers.common.compat.assets import Asset
     from airflow.utils.state import DagRunState, TaskInstanceState
 else:
-    # TODO: Remove this try-exception block after bumping common provider to 1.3.0
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/utils/utils.py:72]: This TODO is probably still needed

File: providers/src/airflow/providers/amazon/aws/assets/s3.py (hunk 1 above)
Comment (potiuk) [reply providers/src/airflow/providers/amazon/aws/assets/s3.py]: Not really. This is a nice by-product of this change.

1) We are not importing from common provider any more, so we will not be bumping it here.
2) When we update min version of provider for airflow, we will automatically fail in the places where the constants were used and they got removed.

For example  - when in the future we will have min_airflow_version = 2.9.0 - we will change the template to remove all _2_9_0_PLUS generated constants. And all the places where those constants were used will automatically fail because the imports will fail (and we will have to fix it).

Comment (potiuk) [reply providers/src/airflow/providers/amazon/aws/assets/s3.py]: I updated the `dev/README_RELEASE_PROVIDER_PACKAGES.md` explainnig (and actually simplifiying and automating detection of such out-dated conditional behaviour) in dev/README_RELEASE_PROVIDER_PACKAGES.md

File: providers/src/airflow/providers/common/io/assets/file.py (hunk 2 above)
Comment (potiuk) [reply providers/src/airflow/providers/common/io/assets/file.py:26]: Ditto.

**Comment by kacpermuda** [comment]
Great idea, thanks Jarek !

File: providers/src/airflow/providers/amazon/aws/assets/s3.py (hunk 1 above)
Comment (potiuk) [reply providers/src/airflow/providers/amazon/aws/assets/s3.py]: Basically - we do not need any relation with common-compat when we write a conditional (`if airflow version`) code in providers - which was really the main complaint (and rightfully so) of @dstandish.

So common-compat should only be used when we have some "real" code in airflow that we need to backport there. Version code is easy to maintain and generate, so it makes sense to generate it in `__init__.py` automatically in those providers that need it (which is precisely what this PR does).

Comment (potiuk) [reply providers/src/airflow/providers/amazon/aws/assets/s3.py]: Let me know @Lee-W @dstandish @uranusjr @eladkal if that approach solves all the concerns (I think I managed to eat cake and have it too in this one, really and made everyone happy - or so I think at least).

Comment (Lee-W) [reply providers/src/airflow/providers/amazon/aws/assets/s3.py]: This one was actually due to a bug related to asset compact in the common provider. are we going to bump the minimum common provider version as well? if not, I think we'll keep the todo till we bump it

**Review by ashb — COMMENTED** [review]
Honestly: This seems crazy magic, complex to understand and massive overkill to do `AIRFLOW_V_3_0_PLUS = Version(Version(AIRFLOW_VERSION).base_version) >= Version("3.0.0")` in a few palces.

-1 to this, it makes things just so much harder to reason about and is not worth the mental gymnastics to it. Sometimes simple duplication is the right answer.

This feels like a very heavyweight solution to a problem that has happened once that I can see.

**Review by dstandish — COMMENTED** [review]
to me, this seems like too much. all we really need is for it to be just a little bit easier to know the core airflow version, it seems to me.

File: providers/src/airflow/providers/amazon/aws/assets/s3.py (hunk 1 above)
Comment (potiuk) [reply providers/src/airflow/providers/amazon/aws/assets/s3.py]: Ah.... I see that was double-compatibility check - but this is not really needed - instead (see the latest push) we should bump the version of compat provider dependency in AWS  (and the compat provider version in provider.yaml) and it should **generally** work as expected - without having to add two compatiblity checks - one in amazon provider and another in the  compat provider. 

I updated it in this way and let's see if it works out-of-the-box.

This is one of the few exceptions where we can bump version of compat provider when we add a new feature to it - then in the dependent provider you bump the minimum version, and since they are going to be released together in the next wave of providers things will work just fine.

**Comment by ashb** [comment]
> to me, this seems like too much. all we really need is for it to be just a little bit easier to know the core airflow version, it seems to me.

Which we already have:

```
AIRFLOW_VERSION = Version(airflow_version)
AIRFLOW_V_3_0_PLUS = Version(AIRFLOW_VERSION.base_version) >= Version("3.0.0")
```

**Comment by ashb** [comment]
There are also changes in this PR like version bumps etc that are entirely unrelated to the PR as given in the description.

**Comment by potiuk** [comment]
Well previous atetmpts led to multiple errors - like updating `__init__.py` that is already auto generated, importing version checks from multiple unrelated places or lack of dependency to a common compat provider where it was needed. 

None of the previous proposals solved all those problems, so I am not sure what is the constructive idea to solve them differently. I would really like to see another, simpler solution for those.

File: providers/src/airflow/providers/common/compat/lineage/hook.py [hunk 5]
```diff
@@ -16,7 +16,10 @@
 # under the License.
 from __future__ import annotations
 
-from airflow.providers.common.compat import AIRFLOW_V_2_10_PLUS, AIRFLOW_V_3_0_PLUS
+from airflow.providers.common.compat import (
+    PROVIDERS_COMMON_COMPAT_AIRFLOW_IS_2_10_PLUS,
+    PROVIDERS_COMMON_COMPAT_AIRFLOW_IS_3_0_PLUS,
```
Comment (ashb) [inline providers/src/airflow/providers/common/compat/lineage/hook.py]: Why did we change the constant name?

**Comment by potiuk** [comment]
> There are also changes in this PR like version bumps etc that are entirely unrelated to the PR as given in the description.

Yep. the changelog was excessive - I just fixed the generateio in latest push and I am going to remove those changes from this PR

File: providers/src/airflow/providers/common/compat/lineage/hook.py (hunk 5 above)
Comment (potiuk) [reply providers/src/airflow/providers/common/compat/lineage/hook.py]: Because this already happened that they were imported from different (and wrong places) - these constants were same named in 3 places and there were few mistakes where they were imported from wrong places.

Comment (potiuk) [reply providers/src/airflow/providers/common/compat/lineage/hook.py]: Many times (and that already happened) people just do "Ctrl-Shift-O" and constants are imported from wrong place and reviewers are missing it (happened already) - renaming those constants makes it much more difficult to make the mistake (for example importing same constant from different provider) when you are reviewing the code.

That's the root cause that already happened in the past and this PR already solved one of those that were left and missed.

Comment (ashb) [reply providers/src/airflow/providers/common/compat/lineage/hook.py]: https://docs.astral.sh/ruff/rules/banned-api/ then.

Comment (potiuk) [reply providers/src/airflow/providers/common/compat/lineage/hook.py]: No. Using this you (unless you find a different way) you have to implement 90+ rules to ban imports from N-1 providers - different for each provider sub tree. Surely we could generate those rules (again by pre-commit probably).

But maybe I am mistaken and there is an easy way to use that ruff rule to ban "every constant named this way from any other package except import of it from my own subtree"

Do you think it is possible @ashb?

**Review by jscheffl — DISMISSED** [review]
Very cool - THANKS!

File: scripts/ci/pre_commit/update_providers_init.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    provider_name = ".".join(provider_folder.relative_to(base_folder).as_posix().split("/"))
+    providers.add(provider_name)
+
+console.print(f"[bright_blue]Regenerating providers __init__ files for providers: {providers}[/]")
+
+if not providers:
+    console.print("[red]\nThe found providers list cannot be empty[/]")
+    sys.exit(1)
+
+res = subprocess.run(
```
Comment (gopidesupavan) [inline scripts/ci/pre_commit/update_providers_init.py]: Please correct me if understand this, When any __init__ file changes, this pre-commit will trigger, compares it with dev/breeze/src/airflow_breeze/templates/PROVIDER__INIT__PY_TEMPLATE.py.jinja2 and the __init__ file.?

When comparison fails instead of displaying an error like `Error while regenerating provider init files`, should we update the error message to indicate that changes to `PROVIDER__INIT__PY_TEMPLATE.py.jinja2` might also be required?

Comment (potiuk) [reply scripts/ci/pre_commit/update_providers_init.py]: Good question @gopidesupavan  It does not trigger comparision but it triggers regeneration of `__init__.py` files - this is generally standard approach of pre-commits - in order to simplify and have auto-fixing capability in pre-commit, usually they work in the way that when they are triggered and it is supposed to produce some generated files it **always** regenerates them.  That is standard convention used in many precommits -  including ours. 

Files are filtered in pre-commit by files/excludes - and it will only run if any of the affected files changed - and it will pass the files as args to the pre-commit scripts (by default), so this is rather efficient - pre-commits are simply skipped unless you modify any of the affect files in your PR.

This has nice effects:

* Pre-commit will automatically "fail" when such generated files are different than the ones alrady in the repository and add the end of pre-commit session it will produce diff of produced files when run with `--show-diff-on-failure` which we do in our CI. 

* When you run it locally, the modified files remain in your local folders in repo - which means that they are changes you can immediately commit - this is basically "auto-fixing" functionality of pre-commit. 

In this case `PROVIDER__INIT__PY_TEMPLATE.py.jinja2` is the "template source" and all the `__init__.py` files should be generated from it, so regeneration of those files means that any change you made locally to those files will be overridden by the content that **should** be there (i.e. generated from the template).

So we do not "require" changes from the template - instead we "require" that all the `__init__.py` files are generated from the template and not modified manually after (which was the root cause why we had #https://github.com/apache/airflow/issues/44024).

**Comment by potiuk** [comment]
After - unnecessary i think - heated discussion I decided to split off (following suggestion of @dstandish) the "non-controversial part of this PR to https://github.com/apache/airflow/pull/44713/ - and will follow up with rebasing/fixing this one after a devlist discussion I am going to start in a moment.

**Comment by potiuk** [comment]
Devlist discussion started https://lists.apache.org/thread/px36w3ph2mf0pmv377dtfc2nhpq8dqw1

File: providers/src/airflow/providers/amazon/aws/assets/s3.py (hunk 1 above)
Comment (potiuk) [reply providers/src/airflow/providers/amazon/aws/assets/s3.py]: I extracted this fix to https://github.com/apache/airflow/pull/44714

**Comment by potiuk** [comment]
> Very cool - THANKS!

Thanks @jscheffl and @kacpermuda  for that, really appreciated.  I was really buffled and I thought I am loosing sanity as I saw very clearly how this one is needed - seeing how strongly @ashb and @dstandish reacted to it. But I see that this is maybe they did not understand the full context and reasoning for the change, or that it is just strong difference in opinions (and that others have different opinions). 

So I extracted all pieces that should be "uncontroversial" from it and started the devlist  discussion above and I hope we can  get to consensus, rebase this PR eventually and maybe apply some comments that might arise separately. I think this is one is pretty useful to have in this form or another, because we already experienced problems with inconsistent version handling.

File: scripts/ci/pre_commit/update_providers_init.py (hunk 6 above)
Comment (gopidesupavan) [reply scripts/ci/pre_commit/update_providers_init.py]: Oh i see , i was in different thought, just compare both and update files if necessary. But this makes sense LGTM :)

**Review by gopidesupavan — DISMISSED** [review]

File: providers/src/airflow/providers/common/compat/lineage/hook.py (hunk 5 above)
Comment (ashb) [reply providers/src/airflow/providers/common/compat/lineage/hook.py]: > That's the root cause that already happened in the past and this PR already solved one of those that were left and missed

The issue that started this change seemed to be about manual changes to generated files, and preventing that is un-contentious. But it them seemed to transform into "lets build a load of scaffolding to be unified about version checks". That's the part I have problems with as this solution makes everything incrementally more complex, each individual change is "yeah okay, slightly complex" but in aggregate all this pre-commit+autogenerated code+magic leads to a hugely complex system.

That said listening to an episode of Talk Python To Me this morning https://import-linter.readthedocs.io/en/stable/index.html came up. I'm trying this quickly to see if we can make it not allow imports from a provider it doesn't list as a dep on.

Comment (ashb) [reply providers/src/airflow/providers/common/compat/lineage/hook.py]: Though as always with new tools: it doesn't like our namespace packages :/

**Comment by ashb** [comment]
A quick POC knocked up using `ruff analyze graph` output:

```
python ./scripts/ci/pre_commit/check_provider_cross_deps.py
warning: `ruff analyze graph` is experimental and may change without warning
'providers/src/airflow/providers/airbyte/hooks/airbyte.py' imports from the 'common.compat' but doesn't list it as a dependency!
'providers/src/airflow/providers/celery/executors/celery_executor_utils.py' imports from the 'standard' but doesn't list it as a dependency!
'providers/src/airflow/providers/edge/example_dags/integration_test.py' imports from the 'standard' but doesn't list it as a dependency!
```

The first was me adding an import to test, the other two are 

```
try:
    from airflow.providers.standard.operators.bash import BashOperator
except ImportError:
    from airflow.operators.bash import BashOperator  # type: ignore[no-redef,attr-defined]
```


<details>

<summary>POC code in here:</summary>

```python
from __future__ import annotations

import json
import os
import subprocess
import sys
from functools import cache
from pathlib import Path

AIRFLOW_ROOT_PATH = Path(__file__).parents[3].resolve()
GENERATED_PROVIDERS_DEPENDENCIES_FILE = AIRFLOW_ROOT_PATH / "generated" / "provider_dependencies.json"
PROVIDER_DEPENDENCIES = json.loads(GENERATED_PROVIDERS_DEPENDENCIES_FILE.read_text())

PROVIDERS_REL_SRC = Path("providers/src/")
PROVIDERS_NS_REL_SRC = Path("providers/src/airflow/providers/")


@cache
def rel_file_to_provider_name(rel_name: str):
    try:
        rel = Path(rel_name).relative_to(PROVIDERS_NS_REL_SRC)

    except ValueError:
        # Not a provider file
        return None

    if rel.suffix == ".py":
        rel = rel.parent

    # This checks for two levels existing as a key in PROVIDER_DEPENDENCIES (i.e. "airbyte" or
    # "common.compat") -- if we ever have an "airflow.providers.ns1.ns2.packge" we will need to change this
    for name in (rel.parts[0], ".".join(rel.parts[0:2])):
        if name in PROVIDER_DEPENDENCIES:
            return name
    return None


def main():
    # Ask ruff to analyze the import graphs
    import_tree_str = subprocess.check_output(
        ["ruff", "analyze", "graph", os.fspath(AIRFLOW_ROOT_PATH / PROVIDERS_REL_SRC)]
    )

    import_tree = json.loads(import_tree_str)
    invalid_imports_found = False

    for file, imports in import_tree.items():
        this_provider = rel_file_to_provider_name(file)
        imported_cross_providers = set(
            filter(
                lambda x: x not in (None, this_provider),
                (rel_file_to_provider_name(imported_file) for imported_file in imports),
            )
        )
        for cross_provider in imported_cross_providers:
            if cross_provider not in PROVIDER_DEPENDENCIES[this_provider]["cross-providers-deps"]:
                # TODO: rich
                print(
                    f"{file!r} imports from the {cross_provider!r} but doesn't list it as a dependency!",
                    file=sys.stderr,
                )
                invalid_imports_found = True


if __name__ == "__main__":
    main()
```

</details>

**Comment by potiuk** [comment]
Interestig. This "ruff analyze graph" is cool. We could use it to replace some of the stuff we do when we generate cross-provider dependencies - we were using it by AST analysis, but we can likely get the speed of rust for that . Really cool tool

**Comment by potiuk** [comment]
OK. I reworked it to follow the suggestion by @ashb (i.e. using `ruff analyze graph`. I managed to work out to work it out to catch both `other provider` imports and `tests_common` wrong import - and added documentation explaining why we are doing it and what the contributors should do, when they want to add version check support to their provider:

Example pre-commit errors:

When you import from tests_common:

image

When you import from another provider:

image

The change is way bigger than the original one, because I had to extract the `version` commands from `tests_utils.compat` to `version_compat` - because `ruff analyze graph`  produces output on a file level, and that was the easy way to check - if `version_compat` is being imported, it should be from the same provider.

Lookig forward to feedback and comments.

**Comment by potiuk** [comment]
BTW. If needs be - I can very easily split that change into extracting the version constants in tests to `version_compat.py` (big but straightforward change) and separately making the "provider version code" usage consistent + pre-commit.

File: providers/src/airflow/providers/common/compat/lineage/hook.py (hunk 5 above)
Comment (potiuk) [reply providers/src/airflow/providers/common/compat/lineage/hook.py]: Yes. Ruff analyze graph seems to do the job nicely.

File: scripts/ci/pre_commit/check_imports_in_providers.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                    console.print(
+                        f"1. Copy `version_compat`.py to `{provider_dir}/version_compat.py` if not there.\n"
+                        f"2. Import the version constants you need as:\n\n"
+                        f"[yellow]from airflow.providers.{provider_id}.version_compat import ...[/]\n"
+                        f"\n"
+                    )
+                    errors_found = True
+
+
+check_imports([AIRFLOW_SOURCES_ROOT_PATH / "providers" / "src", AIRFLOW_SOURCES_ROOT_PATH / "tests_common"])
```
Comment (ashb) [inline scripts/ci/pre_commit/check_imports_in_providers.py:90]: Ruff analyze is quick enough we don't need to worry about only searching files that have changed, right?

**Comment by ashb** [comment]
Thanks very much for making these changes Jarek

File: providers/src/airflow/providers/edge/version_compat.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+from __future__ import annotations
+
+from packaging.version import Version
+
+
+def _get_base_airflow_version() -> Version:
+    from airflow import __version__
+
+    return Version(Version(__version__).base_version)
```
Comment (ashb) [inline providers/src/airflow/providers/edge/version_compat.py]: Any particular reason this is done as a function?

Comment (ashb) [inline providers/src/airflow/providers/edge/version_compat.py]: We don't need base version do we? It will compare fine the way it is won't it?

```suggestion
def _get_base_airflow_version() -> Version:
    from airflow import __version__

    return Version(__version__)
```

File: scripts/ci/pre_commit/check_imports_in_providers.py (hunk 7 above)
Comment (potiuk) [reply scripts/ci/pre_commit/check_imports_in_providers.py:90]: It's not that - this is not about selecting files that changed, but about building the "full" graph of dependencies - including `tests_common'

"ruff analyze" works differently depnding which folders it runs with. I tried a few combos, and sometimes it found dependencies between providers and tests and sometimes they were missing.

For example if you just add "airflow/providers" - it won't find imports to other providers (not sure why - they are missing)

If you specify "airflow/providers/src" as folder - it won't find references to the "tests_common".

So the change here is just a remnant of experimenting with trying to run two ruff commands - but I found that If I pass both - " providers/src" and "tests_common" - both cases of bad imports are found in a single command.

File: providers/src/airflow/providers/edge/version_compat.py (hunk 8 above)
Comment (potiuk) [reply providers/src/airflow/providers/edge/version_compat.py]: Just a bit of reuse - modelled mostly after @dstandish's #44607. We could also inline it as well. Not very tied to it.

Comment (potiuk) [reply providers/src/airflow/providers/edge/version_compat.py]: We do .. This is the main reason why we need to discuss all of that at all. The whole reason why we do this weird `Version(Version.base_version)` is because default comparision rules do not work properly for pre-release version and if we do not do it, checking for features available will break with pre-relase versions. 

I explained it earlier today in https://lists.apache.org/thread/h0s7ypdr05j10jw88b3svj57o4ml1brl - also there is this discussion in PyPI that explains why we have to do it https://github.com/pypa/pip/issues/13089  

Basically it's because of this paragraph in PEP-440:

>  Within a numeric release (1.0, 2.7.3), the following suffixes are
permitted and MUST be ordered as shown:
> .devN, aN, bN, rcN, `<no suffix>`, .postN

This paragraph in PEP 440 and it's implementation in `packaging` means that we have to do this weird looking double version retrieval, because in order to check "is feature available in this version" - we need to basically drop pre-release part of the version (which `Version(Version(__version__).base_version)` does.

Comment (potiuk) [reply providers/src/airflow/providers/edge/version_compat.py]: This is actually the **only** reason we have this discussion at all - anyone else could (like you) made the same mistake and check against `Version(_version_)`. And this is exactly why we have to do hand-holding / guide-railing - because it would break if you do it. For example if you want to do:

```python
from airflow import __version__ 
if Version(__version__) >= Version("3.0.0"):
  # do something only on Airflow 3
```

This condition is `False` on Airflow 3.0.0.dev0, 3.0.0.beta1, 3.0.0.rc1 ... this is the **gist** of the problem we are discussing.

Comment (potiuk) [reply providers/src/airflow/providers/edge/version_compat.py]:
```python
>>> from packaging.version import Version
>>> Version("3.0.0rc1") >= Version("3.0.0")
False
>>> Version(Version("3.0.0rc1").base_version) >= Version("3.0.0")
True
>>>
```

Comment (ashb) [reply providers/src/airflow/providers/edge/version_compat.py]: Ah k! I wonder if we should have a `from airflow import VERSION_INFO` (named to mirror `sys.version_info`) and then we could do `VERSION_INFO >= (3,0,0)`

Doesn't help us for any old checks though. And largely just stylistic.

Comment (potiuk) [reply providers/src/airflow/providers/edge/version_compat.py]: I'd say`_get_base_version` is a better abstraction for it, but yeah. Agree on both accounts - stilistic and does not help for old version checks. 

I can live with both. I just want to make sure for the next 8 months (and beyond) we do not have to deal with this :)

**Review by gopidesupavan — APPROVED** [review]
Nice, thanks jarek for the updates and it looks like `ruff analyze graph` playing nice role.

Great suggestion Ash :)

**Comment by potiuk** [comment]
I think this one should be green. I also extracted out the pure refactor of test version constants to "version_compat" module - that will make this one much smaller (and makes it way easier to see if version_compat is used properly in the `ruff analyze graph` json output.

**Review by jscheffl — APPROVED** [review]

File: providers/src/airflow/providers/edge/version_compat.py (hunk 8 above)
Comment (potiuk) [reply providers/src/airflow/providers/edge/version_compat.py]: Thought about it a bit - and I like the tuple approach more as it does not require top-level imports, which is always a good thing.  

Also this way I think it would satisfy @dstandish approach where he **only** would like to have `get_airflow_version` (or my equivalent `get_base_airflow_version_tuple` method) that could be easily used without using any predefined constants.

Say - you could inline this code then much easier (no additional Version import needed in the provider code):

```
get_airflow_base_version_tuple() > (3, 0, 0) 
```

if you do not want to use the constants (though I think constants are a bit better - mostly because it will be easier to remove them as we move the minimum version of airflow up. One benefit of that could be that every time we add new version of Airflow we woud not have to add a new constant (I think @dstandish  - this is what you really wanted if I read all your past comment and try to extract the real concerns you had).

I propose to merge this one (with the test version_constant  separation in #44770  as prerequisite), and maybe as next step if others  agree - we could change all provider's code to use such inlined version and change instruction on how we should do it). I am happy to make a draft PR for this approach @dstandish - to see how it would look. @ashb  @gopidesupavan @ashb @jscheffl and others - let me know what you think and whether it is worth to do another follow-up - but let's close this one first to not add confusion.

From the very beginning I was not into "let's do it this way" but "let's address the issue we have at hand - for now and the future" - so I am happy to implement it in the way we can all feel comfortable with.  

I know it took a bit of many people's time, but I think it's important to solve it for the next 8 months to come - while I was making this one pass, I just found - out the 7th (!) way that such check was done inside the openlineage (with even differently named constant used to hold airflow version).

Comment (jscheffl) [reply providers/src/airflow/providers/edge/version_compat.py]: I do not have a strong opinion atm and prefer a simple and consolidated approach. Good to merge in my view and also if there is a better solution OK to change it later. Thanks for the energy making it better!

Comment (potiuk) [reply providers/src/airflow/providers/edge/version_compat.py]: Yeah. The way it is changed now, makes it *easy* to change it as a follow-up - no more "common" code impact, dependencies etc., just common convention applied everywhere - and it can be easily changed now as we are guarding it a bit better and telling people what to do to add more compatibility code, so there is a chance both contributors and maintainers will stick to the convention. 

Also - if the #44607 from @dstandish will follow the same (or introduce new - consistently) convention, the pre-commit code we have will guard that the new code added to airflow will not be used until we decide it's OK to use it (basically min-airlfow-version == 2.11) in this case.

Comment (potiuk) [reply providers/src/airflow/providers/edge/version_compat.py]: BTW. I've learned from my friend a very obscene saying in Polish - freely translated `even if sh**tty, it's better  to be consistent` - in Polish it even rhymes ("Lepiej ch**owo ale jednakowo") - which I think applies here very well.

I basically don't really care how we do it, as long as we do it consistently and set the stage to keep it consistent (rather than in 7 different ways).

**Comment by potiuk** [comment]
Rebased after merging #44774  - this should be now much easier to review :)

File: scripts/ci/pre_commit/check_imports_in_providers.py (hunk 7 above)
Comment (potiuk) [reply scripts/ci/pre_commit/check_imports_in_providers.py:90]: And BTW. Yeah - we actually (naturally) get the "selective" per-commit analysis for free anyway. 

The list of filenames affected will be passed anyway to the pre-commit, and this check will only look at the files and will only look for those in the ruff graph.  We could let this pre-commit indeed run it on "all provider files always" - but that would be python and "sligthly" slower to run on all files, but instead we rely on default behaviour of `pre-commit` - which will pass the pre-commit list of changed files (or all files) matching the pre-commit "files:" specification (only provider python files).

In case of CI and `--all-files` - that will be the list of all provider python files passed as arguments

But instead we use "require serial" in pre-commit, which means that we will not run pre-commit in parallell on those files split in chunkse- we only run ruff once - serially, and then we check individually all files passed in Python against the dictionary retrieved.

Should be pretty optimal.

File: providers/src/airflow/providers/edge/version_compat.py (hunk 8 above)
Comment (gopidesupavan) [reply providers/src/airflow/providers/edge/version_compat.py]: I am okay with this changes and to merge, I am also with jens comment , at the moment it is simple and solves our problems.
And i really appreciate your effort :)

Comment (potiuk) [reply providers/src/airflow/providers/edge/version_compat.py]: > And i really appreciate your effort :)

That's my `honey-badger` genes in full swing (cc: @dstandish )

**Comment by potiuk** [comment]
Merging it `as is` then. We can always iterate on it later.
